### ISSUE TITLE
MergeQueue is slow for some query range sets

### ISSUE THREAD

**Issue by lossyrob** [locationtech/geotrellis#2358]
From @mteldridge on Gitter:

> On a different subject, I think I have found an geotrellis issue/bug. I am trying to query a geotrellis layer (hadoop) with a multipolygon made up of about 1600 polys. It looks like MergeQueue isn't up to the task. The layer is indexed with Z3SpaceTimeKey and the query ends up with about 600K index ranges. MergeQueue ends up attempting to insert into the middle of an array 600K times. When it returns, it usually takes >10mins. Often it doesn't return. Perhaps having MergeQueue backed by an array isn't ideal as the query gets large. Any suggestions?
For what its worth, the problem seems to be fixed if the sequence of index ranges is presorted before going into the MergeQueue.
> Offending line is here: https://github.com/locationtech/geotrellis/blob/5c5ebed09f66cde92c77d9f21ce22e65a44573c4/spark/src/main/scala/geotrellis/spark/io/hadoop/formats/FilterMapFileInputFormat.scala#L99

I've experienced similar slowdowns when including time large time ranges against our azavea-datahub Landsat8 2016 layer (used in the FOSS4G 2017 workshop).

If presorting ranges speeds things up, that would be an easy optimization (e.g. sorting here https://github.com/locationtech/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/io/index/MergeQueue.scala#L24). This should be benchmarked and put in if it helps.

**Comment by metasim** [comment]
Do we need #2348 addressed first (or at least the tech stack decided on)? 

Per [@mteldridge's comments](https://gitter.im/geotrellis/geotrellis?at=59b16defee5c9a4c5f42e578), changing the line above to

```scala
val compressedRanges = MergeQueue(r.sortBy(_._1)).toArray
```
significantly improves performance in our use case.

**Comment by mteldridge** [comment]
As @lossyrob mentioned in Gitter, the appropriate fix is to sort by start value in the MergeQueue.apply method prior to inserting the index ranges.  The fix will then get picked up by all the modules that use the MergeQueue.  Pre-sorting should allow one to do the merging in a single pass through the array.   I would also consider replacing the array that backs MergeQueue with something that at least provides constant time insertion/deletion.  This should make the complexity O(nlogn) instead of what it is now - O(n^2) or perhaps worse.

**Comment by mteldridge** [comment]
Here is a candidate for a long-term fix. I get the same answer as the original for our big 1600+ polygon query. It needs some more rigorous testing and profiling though:

```scala
  object MergeQueue {

    def apply(ranges: TraversableOnce[(Long, Long)]): Seq[(Long, Long)] = {

      var merged: List[(Long, Long)] = Nil
      ranges.toSeq.sortBy(_._1).foreach(r => {
        merged = merged match {
          case a :: rest if (r._1 - 1 <= a._2) => (a._1, Math.max(a._2, r._2)) :: rest
          case _ => r :: merged
        }
      })
      merged.reverse
    }
  }
```
Presorting allows a single pass through the data and greatly simplifies the logic.  Should be O(nlogn).

Issues:
- Although the object is binary compatible, the underlying MergeQueue class is gone, which happens to be public, so I guess this is a breaking change.
- Given that this causes a breaking change, I would then change the signature of the `apply` method to:
```scala
    def apply(ranges: SeqLike[(Long, Long)]): Seq[(Long, Long)]
```
which would allow me to call `sortBy` directly on `ranges`
- Needs rigorous testing since it ultimately would go into SFCurve, etc.

If you don't want any breaking changes right now, I would suggest simply doing the presort in the current `MergeQueue.apply` method.

**Comment by mteldridge** [comment]
It looks like Z2 and Z3 internally use on the MergeQueue class directly, so will not benefit from the pre-sort.  The presort still helps significantly in the RDDReader classes.  For now I would suggest simply adding @lossyrob's suggested fix as implemented here:

https://github.com/mteldridge/geotrellis/commit/dc7d7379c174f5bad13b1d727ba913e080a02316

In the long term, one could look into something like this that includes updates to Z2 and Z3.  

https://github.com/mteldridge/geotrellis/commit/51cf29f0575648f1fd00722a22dbc84c9fc3006f

**Comment by mteldridge** [comment]
Also, here is a test that isolates the problem.  The test will hang at the call to `MergeQueue.apply`
https://github.com/locationtech/geotrellis/compare/master...mteldridge:merge-queue-test?expand=1

**Comment by echeipesh** [comment]
Resolved by #2400
